### DISCUSSION TITLE
Biometric scheduled capture not working

### DISCUSSION THREAD

**Discussion by vanyell** [horilla-opensource/horilla#304] [General]
Getting timeout issues

```
[Thu Aug 29 13:51:32.470642 2024] [wsgi:error] [pid 87443:tid 87556] Traceback (most recent call last):
[Thu Aug 29 13:51:32.470674 2024] [wsgi:error] [pid 87443:tid 87556]   File "/home/sysadm/horilla/hris/lib/python3.11/site-packages/zk/base.py", line 249, in __send_command
[Thu Aug 29 13:51:32.470697 2024] [wsgi:error] [pid 87443:tid 87556]     self.__tcp_data_recv = self.__sock.recv(response_size + 8)
[Thu Aug 29 13:51:32.470721 2024] [wsgi:error] [pid 87443:tid 87556]                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[Thu Aug 29 13:51:32.470743 2024] [wsgi:error] [pid 87443:tid 87556] TimeoutError: timed out
[Thu Aug 29 13:51:32.470768 2024] [wsgi:error] [pid 87443:tid 87556] 
[Thu Aug 29 13:51:32.470790 2024] [wsgi:error] [pid 87443:tid 87556] During handling of the above exception, another exception occurred:
[Thu Aug 29 13:51:32.470812 2024] [wsgi:error] [pid 87443:tid 87556] 
[Thu Aug 29 13:51:32.470834 2024] [wsgi:error] [pid 87443:tid 87556] Traceback (most recent call last):
[Thu Aug 29 13:51:32.470857 2024] [wsgi:error] [pid 87443:tid 87556]   File "/home/sysadm/horilla/hris/lib/python3.11/site-packages/apscheduler/executors/base.py", line 125, in run_job
[Thu Aug 29 13:51:32.470880 2024] [wsgi:error] [pid 87443:tid 87556]     retval = job.func(*job.args, **job.kwargs)
[Thu Aug 29 13:51:32.470902 2024] [wsgi:error] [pid 87443:tid 87556]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[Thu Aug 29 13:51:32.470972 2024] [wsgi:error] [pid 87443:tid 87556]   File "/home/sysadm/horilla/biometric/views.py", line 409, in <lambda>
[Thu Aug 29 13:51:32.470999 2024] [wsgi:error] [pid 87443:tid 87556]     lambda: zk_biometric_device_attendance(device.id),
[Thu Aug 29 13:51:32.471050 2024] [wsgi:error] [pid 87443:tid 87556]             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[Thu Aug 29 13:51:32.471075 2024] [wsgi:error] [pid 87443:tid 87556]   File "/home/sysadm/horilla/biometric/views.py", line 1722, in zk_biometric_device_attendance
[Thu Aug 29 13:51:32.471097 2024] [wsgi:error] [pid 87443:tid 87556]     conn.disconnect()
[Thu Aug 29 13:51:32.471120 2024] [wsgi:error] [pid 87443:tid 87556]   File "/home/sysadm/horilla/hris/lib/python3.11/site-packages/zk/base.py", line 394, in disconnect
[Thu Aug 29 13:51:32.471142 2024] [wsgi:error] [pid 87443:tid 87556]     cmd_response = self.__send_command(const.CMD_EXIT)
[Thu Aug 29 13:51:32.471164 2024] [wsgi:error] [pid 87443:tid 87556]                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[Thu Aug 29 13:51:32.471187 2024] [wsgi:error] [pid 87443:tid 87556]   File "/home/sysadm/horilla/hris/lib/python3.11/site-packages/zk/base.py", line 260, in __send_command
[Thu Aug 29 13:51:32.471209 2024] [wsgi:error] [pid 87443:tid 87556]     raise ZKNetworkError(str(e))
[Thu Aug 29 13:51:32.471231 2024] [wsgi:error] [pid 87443:tid 87556] zk.exception.ZKNetworkError: timed out
[Thu Aug 29 13:54:14.831688 2024] [wsgi:error] [pid 87443:tid 87552] Process terminate : timed out
[Thu Aug 29 13:54:27.311127 2024] [wsgi:error] [pid 87443:tid 87556] Process terminate : timed out
```

upon checking the related code

```
def zk_biometric_device_attendance(device_id):
    """
    Retrieve attendance records from a ZK biometric device and update the clock-in/clock-out status.

    :param device_id: The ID of the ZK biometric device.
    """
    device = BiometricDevices.objects.get(id=device_id)
    if device.is_scheduler:
        port_no = device.port
        machine_ip = device.machine_ip
        conn = None
        zk_device = ZK(
            machine_ip,
            port=port_no,
            timeout=5,
            password=0,
            force_udp=False,
            ommit_ping=False,
        )
```

I think timeout should be set to zero following the sample from https://github.com/fananimi/pyzk/blob/master/test_machine.py

editing biometric/views.py... I set the value of timeout to zero and restarted apache2. However I get an error in apache error log...
`[Thu Aug 29 14:18:12.804220 2024] [wsgi:error] [pid 87899:tid 88010] Process terminate : [Errno 11] Resource temporarily unavailable`

This is very confusing, I just changed a single value, 

```
@@ -1669,7 +1669,7 @@ def zk_biometric_device_attendance(device_id):
         zk_device = ZK(
             machine_ip,
             port=port_no,
-            timeout=5,
+            timeout=0,
             password=0,
             force_udp=False,
             ommit_ping=False,
```

**Comment by horilla-opensource** [comment]
> Getting timeout issues
> 
> ```
> [Thu Aug 29 13:51:32.470642 2024] [wsgi:error] [pid 87443:tid 87556] Traceback (most recent call last):
> [Thu Aug 29 13:51:32.470674 2024] [wsgi:error] [pid 87443:tid 87556]   File "/home/sysadm/horilla/hris/lib/python3.11/site-packages/zk/base.py", line 249, in __send_command
> [Thu Aug 29 13:51:32.470697 2024] [wsgi:error] [pid 87443:tid 87556]     self.__tcp_data_recv = self.__sock.recv(response_size + 8)
> [Thu Aug 29 13:51:32.470721 2024] [wsgi:error] [pid 87443:tid 87556]                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> [Thu Aug 29 13:51:32.470743 2024] [wsgi:error] [pid 87443:tid 87556] TimeoutError: timed out
> [Thu Aug 29 13:51:32.470768 2024] [wsgi:error] [pid 87443:tid 87556] 
> [Thu Aug 29 13:51:32.470790 2024] [wsgi:error] [pid 87443:tid 87556] During handling of the above exception, another exception occurred:
> [Thu Aug 29 13:51:32.470812 2024] [wsgi:error] [pid 87443:tid 87556] 
> [Thu Aug 29 13:51:32.470834 2024] [wsgi:error] [pid 87443:tid 87556] Traceback (most recent call last):
> [Thu Aug 29 13:51:32.470857 2024] [wsgi:error] [pid 87443:tid 87556]   File "/home/sysadm/horilla/hris/lib/python3.11/site-packages/apscheduler/executors/base.py", line 125, in run_job
> [Thu Aug 29 13:51:32.470880 2024] [wsgi:error] [pid 87443:tid 87556]     retval = job.func(*job.args, **job.kwargs)
> [Thu Aug 29 13:51:32.470902 2024] [wsgi:error] [pid 87443:tid 87556]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> [Thu Aug 29 13:51:32.470972 2024] [wsgi:error] [pid 87443:tid 87556]   File "/home/sysadm/horilla/biometric/views.py", line 409, in <lambda>
> [Thu Aug 29 13:51:32.470999 2024] [wsgi:error] [pid 87443:tid 87556]     lambda: zk_biometric_device_attendance(device.id),
> [Thu Aug 29 13:51:32.471050 2024] [wsgi:error] [pid 87443:tid 87556]             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> [Thu Aug 29 13:51:32.471075 2024] [wsgi:error] [pid 87443:tid 87556]   File "/home/sysadm/horilla/biometric/views.py", line 1722, in zk_biometric_device_attendance
> [Thu Aug 29 13:51:32.471097 2024] [wsgi:error] [pid 87443:tid 87556]     conn.disconnect()
> [Thu Aug 29 13:51:32.471120 2024] [wsgi:error] [pid 87443:tid 87556]   File "/home/sysadm/horilla/hris/lib/python3.11/site-packages/zk/base.py", line 394, in disconnect
> [Thu Aug 29 13:51:32.471142 2024] [wsgi:error] [pid 87443:tid 87556]     cmd_response = self.__send_command(const.CMD_EXIT)
> [Thu Aug 29 13:51:32.471164 2024] [wsgi:error] [pid 87443:tid 87556]                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> [Thu Aug 29 13:51:32.471187 2024] [wsgi:error] [pid 87443:tid 87556]   File "/home/sysadm/horilla/hris/lib/python3.11/site-packages/zk/base.py", line 260, in __send_command
> [Thu Aug 29 13:51:32.471209 2024] [wsgi:error] [pid 87443:tid 87556]     raise ZKNetworkError(str(e))
> [Thu Aug 29 13:51:32.471231 2024] [wsgi:error] [pid 87443:tid 87556] zk.exception.ZKNetworkError: timed out
> [Thu Aug 29 13:54:14.831688 2024] [wsgi:error] [pid 87443:tid 87552] Process terminate : timed out
> [Thu Aug 29 13:54:27.311127 2024] [wsgi:error] [pid 87443:tid 87556] Process terminate : timed out
> ```
> 
> upon checking the related code
> 
> ```
> def zk_biometric_device_attendance(device_id):
>     """
>     Retrieve attendance records from a ZK biometric device and update the clock-in/clock-out status.
> 
>     :param device_id: The ID of the ZK biometric device.
>     """
>     device = BiometricDevices.objects.get(id=device_id)
>     if device.is_scheduler:
>         port_no = device.port
>         machine_ip = device.machine_ip
>         conn = None
>         zk_device = ZK(
>             machine_ip,
>             port=port_no,
>             timeout=5,
>             password=0,
>             force_udp=False,
>             ommit_ping=False,
>         )
> ```
> 
> I think timeout should be set to zero following the sample from https://github.com/fananimi/pyzk/blob/master/test_machine.py
> 
> editing biometric/views.py... I set the value of timeout to zero and restarted apache2. However I get an error in apache error log... `[Thu Aug 29 14:18:12.804220 2024] [wsgi:error] [pid 87899:tid 88010] Process terminate : [Errno 11] Resource temporarily unavailable`
> 
> This is very confusing, I just changed a single value,
> 
> ```
> @@ -1669,7 +1669,7 @@ def zk_biometric_device_attendance(device_id):
>          zk_device = ZK(
>              machine_ip,
>              port=port_no,
> -            timeout=5,
> +            timeout=0,
>              password=0,
>              force_udp=False,
>              ommit_ping=False,
> ```

HI @vanyell ,
Which ZK device are you using ?

**Reply by vanyell** [reply]
ZK Multibio 800

**Reply by horilla-opensource** [reply]
Was the test connection successful?

**Reply by vanyell** [reply]
yes the test connection was successful
running the script from fananimi's repo.. also works

**Reply by vanyell** [reply]
Still having this nagging problem. Made a few changes to the code below.

```
diff --git a/biometric/views.py b/biometric/views.py
index a9a490b2..2434ab80 100644
--- a/biometric/views.py
+++ b/biometric/views.py
@@ -1669,15 +1669,17 @@ def zk_biometric_device_attendance(device_id):
         zk_device = ZK(
             machine_ip,
             port=port_no,
-            timeout=5,
+            timeout=0,
             password=0,
             force_udp=False,
             ommit_ping=False,
+            verbose=True,
         )
         try:
             conn = zk_device.connect()
-            conn.enable_device()
+            conn.disable_device()
             attendances = conn.get_attendance()
+            conn.enable_device()
             last_attendance_datetime = attendances[-1].timestamp
             if device.last_fetch_date and device.last_fetch_time:
```
Still getting 
`[Fri Aug 30 19:33:06.940941 2024] [wsgi:error] [pid 14095:tid 14203] Process terminate : [Errno 11] Resource temporarily unavailable`

This is really puzzling.

**Comment by vanyell** [comment]
Just an observation.... when using Horilla for integrating with biometric devices for time keeping, should the device be used solely for time keeping and not for door access?

**Reply by horilla-opensource** [reply]
conn.unlock() using this part of line we can access the door . 

However, this functionality is not currently included in the default code to ensure that the system is used primarily for timekeeping, as per standard configurations.                                                                                                                                     
If you prefer to use the biometric device for door access as well, we can add the conn.unlock() line to the def biometric_device_test(request, device_id):   function in Horilla Source Code.                                                                             
Once this change is made, you will be able to unlock the door by simply clicking the button labeled "Test" within the Horilla interface

**Reply by vanyell** [reply]
no worries. I'll test it out first... no need to add into the code base.
